packages/utils/telemetry-utils/src/errorLoggingInternalHelpers.ts

```diff
@@ -0,0 +1,94 @@
+export const isRegularObject = (value: any): boolean => {
+    return value !== null && !Array.isArray(value) && typeof value === "object";
+};
+
+/** Inspect the given error for common "safe" props and return them */
+export function extractLogSafeErrorProperties(error: any) {
+    const removeMessageFromStack = (stack: string, errorName?: string) => {
+        const stackFrames = stack.split("\n");
+        stackFrames.shift(); // Remove "[ErrorName]: [ErrorMessage]"
+        if (errorName !== undefined) {
+            stackFrames.unshift(errorName); // Add "[ErrorName]"
+        }
+        return stackFrames.join("\n");
+    };
+
+    const message = (typeof error?.message === "string")
+        ? error.message as string
+        : String(error);
+
+    const safeProps: { message: string; errorType?: string; stack?: string } = {
+        message,
+    };
+
+    if (isRegularObject(error)) {
+        const { errorType, stack, name } = error;
+
+        if (typeof errorType === "string") {
+            safeProps.errorType = errorType;
+        }
+
+        if (typeof stack === "string") {
+            const errorName = (typeof name === "string") ? name : undefined;
+            safeProps.stack = removeMessageFromStack(stack, errorName);
+        }
+    }
+
+    return safeProps;
+}
+
+/** Copy props from source onto target, overwriting any keys that are already set on target */
+export function copyProps(target: unknown, source: ITelemetryProperties) {
+    Object.assign(target, source);
+}
+
+/**
+ * Read-Write Logging Error.  Not exported.
+ * This type alias includes addTelemetryProperties, and applies to objects even if not instanceof LoggingError\
+ */
+type RwLoggingError = LoggingError;
+
+/** Type guard for RwLoggingError.  Not exported. */
+const isRwLoggingError = (x: any): x is RwLoggingError =>
+    typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
+
+/**
+ * Mixes in the given properties to the given error object, then accessible via ILoggingError.getTelemetryProperties
+ * @param errorObject - An object (MUST be non-null non-array object type, not frozen) to add telemetry features to
+ * @param props - The telemetry props to add to the errorObject
+ * @returns the same object that was passed in, now also implementing ILoggingError.
+ */
+export function mixinTelemetryProps<T>(
+    errorObject: T,
+    props: ITelemetryProperties,
+): asserts errorObject is T & ILoggingError {
+    assert(isRegularObject(errorObject) && !Object.isFrozen(errorObject), "Cannot mixin Telemetry Props");
+
+    if (isRwLoggingError(errorObject)) {
+        errorObject.addTelemetryProperties(props);
+        return;
+    }
+
+    // Even though it's not exposed, fully implement RwLoggingError for subsequent calls to mixinTelemetryProps
+    const loggingError = errorObject as T & RwLoggingError;
+
+    const propsForError = {...props};
+    loggingError.getTelemetryProperties = () => propsForError;
```

Comment:
So after that operation, loggingError.getTelemetryProperties().errorType !== undefined, but loggingError.errorType === undefined, right?
And going through mixinTelemetryPropsWithFluidError() cycle one more type (i.e. calling annotate again) would lose such errorType, I believe.

If that's the case, that's an example of duality that I do not like - it almost feels like real logging object, but it's not and various code paths would lead to unpredictable results.

Reply:
Take a look at the current change and let me know if you still have concerns.  I never quite saw the bug here.

Reply:
> object = { errorType: "foo", fluidErrorCode: "bar", x: 10 }
object2 = normalizeError(object, { x: 5, errorType: "foobar" });

I'd expect these things to be true:

> object2.errorType === object2.getTelemetryProperties().errorType

That's not the case as mixinTelemetryPropsWithFluidError does not let to overwrite errorType (it drops it).


> object2.x === object2.getTelemetryProperties().x

That's not true either, as there is no connection between the two.


> object2.errorType = "lake";

I'd expect this statement to change what is recorded in telemetry. It will be ignored, prior value (accessible through object2.getTelemetryProperties().errorType) will be recorded in telemetry. Same for .x property.

There is this duality that is hard to grasp unless you know the inner details and know that you should not touch original object. This is mostly true if developer controls initial creation of object, so such developer knows its content, but decided to use normalizeError() to change it (maybe one layer down the stack) and gets into a state that is much harder to reason about.
If we want to go that route, I'd suggest moving toward getTelemetryProperties() only returning a list of property names that are Ok to record in telemetry, and actual content coming from object itself, such that there was single source of truth. That way a process of annotation is much more incremental and less invasive.

Reply:
Ok.  We just have different expectations.  I see the telemetry props functions as shortcuts to include stuff to log alongside the error.  You see the telemetry props functions as inherently accessing the object itself.  I'm not really worried about someone overwriting `errorType` in telemetry props.  Why would they? But you're right that it's possible.

We can discuss more in #6720... :)

Reply:
> know that you should not touch original object

I think this is good coding practice anyway.  Only access objects according to their type.  Otherwise you're off in indeterminate land as far as I'm concerned.